src/Processors/Executors/PipelineExecutor.cpp

```diff
@@ -364,13 +364,12 @@ void PipelineExecutor::initializeExecution(size_t num_threads, bool concurrency_
         /// To avoid counting them in ConcurrencyControl, we create dummy slot allocation.
         cpu_slots = grantSlots(num_threads);
     }
-    size_t use_threads = cpu_slots->grantedCount();
 
     Queue queue;
     Queue async_queue;
     graph->initializeExecution(queue, async_queue);
 
-    tasks.init(num_threads, use_threads, profile_processors, trace_processors, read_progress_callback.get());
+    tasks.init(num_threads, 1, profile_processors, trace_processors, read_progress_callback.get());
```

Comment:
I don't know where exactly (maybe here) but please document somewhere the purpose of the optimization, i.e. why the thread count is initially 1 and incremented incrementally afterwards. Otherwise, readers of this code will need to do Git archeology.

Reply:
Because one of the key conditions to determine whether tasks are finished is num of `use_threads` threads all going into idle (placed into `threads_queue`). I think this variable should somewhat reflect number of spawned threads, otherwise, we need to wait for all threads spawned even if these is no job to do. 

Setting to 1 instead of 0 is to tackle the problem in single thread. 

Will add the doc.